zabbix: adding zabbix proxy support

mmahut opened the pull request:
###### Motivation for this change

Adding zabbix proxy support by creating a new zabbix proxy module along with zabbix-proxy subpackages.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [X] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

mmahut committed:
zabbix: adding proxy support

mmahut committed:
zabbix20: adding proxy support

mmahut committed:
zabbix22: adding proxy support

mmahut committed:
zabbix34: adding proxy support

mmahut committed:
modules: adding zabbix-proxy

andir commented:
@GrahamcOfBorg build zabbix zabbix20 zabbix34 zabbix22

Can you maybe write a test for the module?

mmahut committed:
nixosTests.zabbix-proxy: init

mmahut commented:
@andir not a bad idea at all, the service is fairly simple. Added the test.

aanderse commented on nixos/modules/services/monitoring/zabbix-proxy.nix:
```diff
@@ -0,0 +1,133 @@
+       '';
+    };
+
+    services.zabbixProxy.dbServer = mkOption {
```
My own opinion here... feel free to take it or leave it.

In https://github.com/NixOS/nixpkgs/issues/62073 I'm trying to standardize database naming conventions and options available. It would be awesome if in this PR you added a few more options like `type`, `port`, `user`, `socket`, and `createLocally`. You could take a look at https://github.com/NixOS/nixpkgs/commit/615f8b8982b26bbb1a3e202be020d27a9f205c62 to see a decent example of what I'm thinking of.

Again... my opinion, take it or leave it. :smile:

aanderse commented on nixos/modules/services/monitoring/zabbix-proxy.nix:
```diff
@@ -0,0 +1,133 @@
+
+        path = [ pkgs.nettools ];
+
+        serviceConfig.ExecStart = "@${pkgs.zabbix.proxy}/sbin/zabbix_proxy zabbix_proxy --config ${configFile}";
```
Style nitpick but it is preferred to avoid extra typing like so:
```
serviceConfig = {
  ExecStart = "@${pkgs.zabbix.proxy}/sbin/zabbix_proxy zabbix_proxy --config ${configFile}";
  Type = "forking";
  ...
};
```

aanderse commented on nixos/modules/services/monitoring/zabbix-proxy.nix:
```diff
@@ -0,0 +1,133 @@
+    ''
+      Server = ${cfg.server}
+
+      LogFile = ${logDir}/zabbix_proxy
```
Logging to the system log is preferred if possible. If not possible does `zabbix` take care of its own log rotation?

mmahut replied:
Sadly no systemd integration here. We will use `logrotate`.

aanderse commented on nixos/modules/services/monitoring/zabbix-proxy.nix:
```diff
@@ -0,0 +1,133 @@
+
+  ###### interface
+
+  options = {
```
Less typing preferred, instead of writing `services.zabbixProxy.*` over and over you can write `services.zabbixProxy = { ... };`

aanderse reviewed (changes requested):
Please let me know if you have any questions or need a hand with anything.

mmahut commented on pkgs/servers/monitoring/zabbix/2.0.nix:
```diff
@@ -89,4 +89,31 @@ in
+      description = "An enterprise-class open source distributed monitoring solution (client-server proxy)";
+      homepage = https://www.zabbix.com/;
+      license = licenses.gpl2;
+      maintainers = [ maintainers.eelco ];
```
I added mysel as a maintainer.

mmahut commented on nixos/modules/services/monitoring/zabbix-proxy.nix:
```diff
@@ -0,0 +1,133 @@
+      '';
+    };
+
+    services.zabbixProxy.dbPassword = mkOption {
```
What is the best way to paste the content of the file into `zabbix_proxy.conf` which is writen using `configFile`? I haven't find anything useful.

mmahut commented on nixos/modules/services/monitoring/zabbix-proxy.nix:
```diff
@@ -0,0 +1,133 @@
+        description = "Zabbix daemon user";
+      };
+
+    systemd.services."zabbix-proxy" =
```
What is the best way to be able to run `sudo` without using `PermissionsStartsOnly`?

aanderse replied:
Does this service require running as `root`? Or is it the `postgresql` stuff that requires `root`/`sudo`?

mmahut replied:
I need [sudo](url) to `postgresql` user in order to create the database and its associated user.

aanderse replied:
I mentioned `services.postgresql.ensureDatabases` and `services.postgresql.ensureUsers` in another comment as they can be used to provision a database. Specifically take a look at https://github.com/NixOS/nixpkgs/pull/61923/commits/615f8b8982b26bbb1a3e202be020d27a9f205c62 post to see how I replaced usage of `sudo` to provision a `postgresql` database.

mmahut committed:
zabbix: ecaping homepage and adding myself as a maintainer

mmahut committed:
modules: various fixes to the zabbix proxy module

- not running the proxy under root
- logrotate
- adding more database configuration options
- adding support for password file

mmahut commented:
@GrahamcOfBorg eval

aanderse commented:
I was looking into the configuration on the zabbix site and realized that most the versions we ship are badly out of date and unsupported upstream. We shouldn't be shipping software that isn't supported upstream.

mmahut commented:
You are right. I plan to do a major upgrade and switch to 3.4 as a default.
But that is a major non compatible change that will have to go to release
notes.

But that is for another pull request :)

On Sun, Jun 16, 2019, 02:10 Aaron Andersen <notifications@github.com> wrote:

> I was looking into the configuration on the zabbix site and realized that
> most the versions we ship are badly out of date and unsupported upstream.
> We shouldn't be shipping software that isn't supported upstream.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/63130?email_source=notifications&email_token=AAAZSW6UAS3J54LFA2DVPX3P2WAI5A5CNFSM4HYIZ7OKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODXZCLZI#issuecomment-502408677>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAZSW7PBEJ5A5KMQHXT3FTP2WAI5ANCNFSM4HYIZ7OA>
> .
>

mmahut commented:
Implemented in https://github.com/NixOS/nixpkgs/pull/63844
